### PR TITLE
channels: Fix bad initialization of RNG seeds.

### PR DESCRIPTION
**Pull request by MattEttus**
---
name: channels: Fix bad initialization of RNG seeds.
about: 
title: ''
labels: ''
assignees: ''

---

# Pull Request Details

## Description
Previosuly the block attempted to initialize RNGs in the constructor initializers, but this does not work.  Moved the initialization to the body of the constructor.

## Checklist

- [X] I have read the [CONTRIBUTING document](https://github.com/gnuradio/gnuradio/blob/master/CONTRIBUTING.md).
- [ ] I have squashed my commits to have one significant change per commit. 
- [X] I [have signed my commits before making this PR](https://github.com/gnuradio/gnuradio/blob/master/CONTRIBUTING.md#dco-signed)
- [X] My code follows the code style of this project. See [GREP1.md](https://github.com/gnuradio/greps/blob/master/grep-0001-coding-guidelines.md).
- [ ] I have updated [the documentation](https://wiki.gnuradio.org/index.php/Main_Page#Documentation) where necessary.
- [ ] I have added tests to cover my changes, and all previous tests pass.

### CODE REVIEW

File: gr-channels/lib/flat_fader_impl.cc [hunk 1]
```diff
@@ -38,6 +35,12 @@ flat_fader_impl::flat_fader_impl(uint32_t N, float fDTs, bool LOS, float K, uint
       scale_los(sqrtf(d_K) / sqrtf(d_K + 1)),
       scale_nlos(1 / sqrtf(d_K + 1))
 {
+    d_theta = dist_1(rng_1);
```
Comment (willcode) [inline gr-channels/lib/flat_fader_impl.cc:38]: rng_1 is used here but is set below. Is that intentional?

Comment (MattEttus) [reply gr-channels/lib/flat_fader_impl.cc:41]: yes, my mistake, I have pushed a fix to reverse it.

**Review by willcode — APPROVED** [review]

**Comment by willcode** [comment]
Backport note: 3.8 uses boost instead of std, and the code is a little different.

**Comment by marcusmueller** [comment]
Debian 11 failure is due to a volk bug, not a blocker here

**Comment by marcusmueller** [comment]
> Backport note: 3.8 uses boost instead of std, and the code is a little different.

Ha! We've got xoroshiro128p.h so that we depend on neither boost nor std::mt19937 (both being slow) (and originally, because there were many places where people used `rand` / `srand` to generate long sequences (ouch) of uniform (double-ouch) random variables in multi-threaded GNU Radio (pain) from back in the STS days, and that had to be replaced with something).

I'm not saying we need to replace `<random>`/`<boost/random.h>` mt19937 with `"gnuradio/xorohsiro128p.h"`, as that leaves the problem of how to generate anything but uniform 64 bit ints as an exercise for the reader. Just a heads up that we do have a faster random engine if this ever becomes a bottleneck.

**Comment by marcusmueller** [comment]
oops: 

> I'm not saying we need to replace `<random>`/`<boost/random.h>` mt19937 with `"gnuradio/xorohsiro128p.h"`, as that leaves the problem of how to generate anything but uniform 64 bit ints as an exercise for the reader. Just a heads up that we do have a faster random engine if this ever becomes a bottleneck.

I forgot I've implemented `gr::xoroshiro128p_prng`, which works exactly the same as `std::mt19937`, but again, not a strong recommendation to use that: it's more performant, that's all.

**Review by mormj — APPROVED** [review]
LGTM

**Comment by willcode** [comment]
The v3.8 code has the same problem. Code was restructured since then, so cherry picking won't apply. Anyone want to take a hack at it?

**Comment by willcode** [comment]
Skipping backport to 3.8. If someone wants to do the backport and test, we can revisit.

**Pull request merged**